core/src/main/java/org/apache/iceberg/hadoop/HadoopStreams.java

```diff
@@ -185,8 +185,21 @@ public void flush() throws IOException {
 
     @Override
     public void close() throws IOException {
-      stream.close();
-      this.closed = true;
+      try {
+        stream.close();
+        // {@link org.apache.hadoop.fs.s3a.S3ABlockOutputStream#close()} calls {@link
+        // org.apache.hadoop.fs.s3a.S3ABlockOutputStream#putObject()}
+        // which doesn't throw an exception when interrupted. Need to check the interrupted flag to
+        // detect failed object upload
+        // and propagate the error up.
+        if (Thread.interrupted()
+            && "org.apache.hadoop.fs.s3a.S3ABlockOutputStream"
+                .equals(stream.getWrappedStream().getClass().getName())) {
+          throw new IOException("object upload failed as it was interrupted during close");
+        }
+      } finally {
+        this.closed = true;
```

Comment:
why not just set this right after `stream.close()`? Then there wouldn't be a need to have the `finally` block

Reply:
`stream.close` can throw an exception. we want to mark this closed no matter what that's why we need the finally block here.

Reply:
if we want to set `this.closed = true` no matter the outcome, then we could also set it right before calling close

Reply:
sure I can do that. but that's what `finally` block does as well by marking it to true _no matter what_.

Trying to understand better if its not recommended to use finally or it has any drawbacks.

Reply:
Personally I am slightly favoring changing `closed` after the stream is actually closed. I checked the current style in Iceberg. I saw two styles: (1) setting `closed` before the action (2) just set it after the action without finally. It is not be necessary to set it In error/failure case, because close didn't complete successfully.

Reply:
> Personally I am slightly favoring changing `closed` after the stream is actually closed. I checked the existing Iceberg code on handling close. I saw two styles: (1) setting `closed` before the action (2) just set it after the action without finally. It is not be necessary to set it In error/failure case, because close didn't complete successfully.

yes that was my observation as well when looking through the codebase, meaning we can probably have 
```
public void close() throws IOException {
    stream.close();
    this.closed = true;
    if (interrupted && ...) {
        throw IOException(..);
    }
}
```
without needed the try-finally block

Reply:
ok, removed the finally block.